h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java

```diff
@@ -309,13 +310,25 @@ private ValFrame frame_op_frame(Frame lf, Frame rt) {
       } else
         throw new IllegalArgumentException("Frames must have same rows, found " + lf.numRows() + " rows and " + rt.numRows() + " rows.");
     }
+
+    final int[][] alignedCategoricals = new int[lf.numCols()][];
+
+    for (int c = 0; c < lf.numCols(); c++) {
+      if (lf.vec(c).isCategorical() && rt.vec(c).isCategorical()) {
+        if (lf.vec(c).domain().length >= rt.vec(c).domain().length) {
```

Comment:
Pass arguments in right order beforehand to avoid swaping. For high count of domain levels, this may save some memory for a short period of time.